src/main/java/org/jenkinsci/main/modules/instance_identity/InstanceIdentity.java

```diff
@@ -59,6 +59,12 @@ public InstanceIdentity(File keyFile) throws IOException {
             try {
                 in = new StringReader(new String(KEY.decrypt().doFinal(enc), "UTF-8"));
             } catch (GeneralSecurityException x) {
+                LOGGER.log(Level.SEVERE, "identity.key.enc is corrupted. Deleting identity.key.enc and generating a new one");
+                Util.deleteFile(keyFile);
+                gen.initialize(2048, new SecureRandom()); // going beyond 2048 requires crypto extension
+                keys = gen.generateKeyPair();
+                write(keys, keyFile);
+                LOGGER.log(Level.SEVERE, "New identity.key.enc was successfully created");
                 throw new IOException(x);
```

Comment:
I thought the whole point was to recover gracefully! So then you should log this exception (on line 62), and continue without the `read`.

Also here you are duplicating the original code on lines 66–68, which is an indication that the whole if-elseif-else block needs to be restructured a bit:

```java
if (oldKeyFile.exists()) {
    // …
} else {
    // if new file exists, and can be read, load it, else generate new
}
```